src/v/cluster/cluster_epoch_service.h

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "cluster/partition_leaders_table.h"
+#include "raft/consensus.h"
+#include "raft/group_manager.h"
+#include "rpc/connection_cache.h"
+#include "ssx/checkpoint_mutex.h"
+
+#include <seastar/core/distributed.hh>
+#include <seastar/core/gate.hh>
+
+namespace cluster {
+
+class controller_stm;
+
+// A sharded service on every core that is responsible for returning the
+// cluster's epoch.
+//
+// The cluster epoch is a monotonically increasing value that is currently used
+// in the Cloud Topics's L0 implementation.
+class cluster_epoch_service
```

Comment:
nit: may be add some detail on how the epochs are propagated to shards (some high level design comment).